src/nix/narinfo.rs

```diff
@@ -0,0 +1,246 @@
+#[derive(
+    Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize, schemars::JsonSchema,
+)]
+
+/// `NarInfo` is a struct that represents the information about a Nix Archive (nar).
+pub struct NarInfo {
+    /// The path to the store where the nar is located.
+    pub store_path: String,
+    /// The URL where the nar can be downloaded.
+    pub url: String,
+    /// The compression method used on the nar.
+    pub compression: String,
+    /// The hash of the file.
+    pub file_hash: String,
+    /// The size of the file in bytes.
+    pub file_size: usize,
+    /// The hash of the nar.
+    pub nar_hash: String,
+    /// The size of the nar in bytes.
+    pub nar_size: usize,
+    /// The deriver of the nar, if any.
+    pub deriver: Option<String>,
+    /// The system for which the nar is intended, if specified.
+    pub system: Option<String>,
+    /// The references of the nar.
+    pub references: Vec<String>,
+    /// The signature of the nar.
+    pub sig: String,
+    /// The content addressable storage identifier of the nar, if any.
+    pub ca: Option<String>,
+}
+
+impl NarInfo {
+    /// Fetches the narinfo file for a given output path from a list of servers.
+    ///
+    /// # Arguments
+    ///
+    /// * `output_path` - The output path of the narinfo file.
+    /// * `servers` - A slice of server URLs to fetch the narinfo file from.
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result` containing an `Option` of `NarInfo` if the narinfo file is found, or `None` if it is not found.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the output path is invalid or if there is an error fetching or parsing the narinfo file.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use nixtract::narinfo::NarInfo;
+    ///
+    /// let output_path = "/nix/store/abc123";
+    /// let servers = vec!["server1.example.com".to_string(), "server2.example.com".to_string()];
+    ///
+    /// match NarInfo::fetch(output_path, &servers) {
+    ///     Ok(Some(narinfo)) => {
+    ///         // Narinfo file found, do something with it
+    ///     },
+    ///     Ok(None) => {
+    ///         // Narinfo file not found
+    ///     },
+    ///     Err(err) => {
+    ///         // Error fetching or parsing narinfo file
+    ///     }
+    /// }
+    /// ```
+    pub fn fetch(output_path: &str, servers: &[String]) -> crate::error::Result<Option<Self>> {
```

Comment:
That's probably too costly (and if not you'll hate me for suggesting it that late), but mentioning it any way for completeness: running `nix path-info --json --store {binaryCacheUrl}` can get you the store path information in json, and for arbitrary stores (including the local one, which can come in handy I guess)

Reply:
The current approach prevents me from having to go out of Rust and spawn another process. I will leave it as it is, until we encounter an issue with the current approach.